rastervision_pytorch_learner/rastervision/pytorch_learner/learner.py

```diff
@@ -799,23 +799,19 @@ def _predict_dataloader(
                 might or might not be batched depending on the batched_output
                 argument.
         """
-        model_train_state = self.model.training
         self.model.eval()
 
-        with torch.inference_mode():
-            for x, y in dl:
-                x = self.to_device(x, self.device)
-                z = self.predict(x, raw_out=raw_out, **predict_kw)
-                x = self.to_device(x, 'cpu')
-                y = self.to_device(y, 'cpu') if y is not None else y
-                z = self.to_device(z, 'cpu')
-                if batched_output:
-                    yield x, y, z
-                else:
-                    for _x, _y, _z in zip(x, y, z):
-                        yield _x, _y, _z
-
-        self.model.train(model_train_state)
```

Comment:
>We also get rid of the call to self.model.train because it will only get called if all the items are iterated through

This is unfortunate. Now this method ends up having a side-effect. Can you add a small note to the docstring that says that this will put the model in eval mode?

Reply:
Yes

Reply:
Actually, this might be unnecessary since the `predict` method is getting called and we could just put the model into training mode at the end of that. That could have some performance implications though. I'm going to look into it.

Reply:
The model is put into train and eval mode in many other places in `Learner` and no attempt is made to put it back into the mode it was in before, so this change actually makes it more consistent with the rest of the class. This also mimics the behavior of PyTorch. There was some [discussion](https://discuss.pytorch.org/t/opinion-eval-should-be-a-context-manager/18998) about making `eval()` a context manager the same way `inference_mode` is. I'd rather just leave things as they are. I'm not opposed to changing the documentation to state that this is the behavior, but many methods will need to have this warning. 

Another issue is that the `predict` and `validate_step` methods never call `model.eval()` and are depending on the caller to do this which is dangerous. These methods should call `model.eval` themselves in case they are called directly.

I ran a little experiment and the time it takes to put a model into eval mode and then back into training mode is about 1/20 the time it takes to run a typical batch (of shape [32, 3, 256, 256]) through a Resnet18 when running on a Colab GPU. So the overhead of putting the model into eval mode for each batch is non-negligible, but not huge, and would make things safer.

Reply:
Thanks for looking into this. I think doing it per batch is a bit much. PyTorch Lightning also seems [to do it only once at the start of training](https://github.com/Lightning-AI/lightning/blob/6b0b6b8903f3931f66360ac7ce0cd9e4d04e3f84/src/pytorch_lightning/trainer/trainer.py#L1154). We can leave it as it is for now.